// @unovis/vue/components/nested-donut/index.js
import { defineComponent as m, inject as d, computed as y, ref as g, onMounted as v, nextTick as f, onUnmounted as b, watch as u, openBlock as S, createElementBlock as B } from "vue";
import { NestedDonut as c } from "@unovis/ts";
import { useForwardProps as _, arePropsEqual as h } from "../../utils/props.js";
import { componentAccessorKey as w } from "../../utils/context.js";
const F = { "data-vis-component": "" }, C = c.selectors, D = /* @__PURE__ */ m({
  __name: "index",
  props: {
    angleRange: {},
    direction: {},
    value: { type: [Function, Number, null] },
    centralLabel: {},
    centralSubLabel: {},
    centralSubLabelWrap: { type: Boolean },
    showBackground: { type: Boolean },
    sort: { type: Function },
    layers: {},
    layerSettings: { type: [Function, null] },
    layerPadding: {},
    cornerRadius: {},
    emptySegmentAngle: {},
    hideOverflowingSegmentLabels: { type: Boolean },
    segmentColor: { type: [Function, String, Array, null] },
    segmentLabel: { type: [Function, String, null] },
    segmentLabelColor: { type: [Function, String, Array, null] },
    showEmptySegments: { type: Boolean },
    showSegmentLabels: { type: Boolean },
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(i, { expose: p }) {
    const r = i, n = d(w), o = y(() => n.data.value ?? r.data), a = _(r), t = g();
    return v(() => {
      f(() => {
        var e;
        t.value = new c(a.value), (e = t.value) == null || e.setData(o.value), n.update(t.value);
      });
    }), b(() => {
      var e;
      (e = t.value) == null || e.destroy(), n.destroy();
    }), u(a, (e, l) => {
      var s;
      h(e, l) || (s = t.value) == null || s.setConfig(a.value);
    }), u(o, () => {
      var e;
      (e = t.value) == null || e.setData(o.value);
    }), p({
      component: t
    }), (e, l) => (S(), B("div", F));
  }
});
export {
  C as VisNestedDonutSelectors,
  D as default
};
