// @unovis/vue/components/donut/index.js
import { defineComponent as d, inject as m, computed as v, ref as g, onMounted as y, nextTick as f, onUnmounted as b, watch as u, openBlock as _, createElementBlock as h } from "vue";
import { Donut as l } from "@unovis/ts";
import { useForwardProps as B, arePropsEqual as F } from "../../utils/props.js";
import { componentAccessorKey as S } from "../../utils/context.js";
const k = { "data-vis-component": "" }, E = l.selectors, L = /* @__PURE__ */ d({
  __name: "index",
  props: {
    id: { type: Function },
    value: { type: [Function, Number, null] },
    angleRange: {},
    padAngle: {},
    sortFunction: { type: Function },
    cornerRadius: {},
    color: { type: [Function, String, Array, null] },
    radius: {},
    arcWidth: {},
    centralLabel: {},
    centralSubLabel: {},
    centralSubLabelWrap: { type: Boolean },
    showEmptySegments: { type: Boolean },
    emptySegmentAngle: {},
    showBackground: { type: Boolean },
    backgroundAngleRange: {},
    duration: {},
    events: {},
    attributes: {},
    data: {}
  },
  setup(p, { expose: i }) {
    const r = p, o = m(S), n = v(() => o.data.value ?? r.data), a = B(r), t = g();
    return y(() => {
      f(() => {
        var e;
        t.value = new l(a.value), (e = t.value) == null || e.setData(n.value), o.update(t.value);
      });
    }), b(() => {
      var e;
      (e = t.value) == null || e.destroy(), o.destroy();
    }), u(a, (e, c) => {
      var s;
      F(e, c) || (s = t.value) == null || s.setConfig(a.value);
    }), u(n, () => {
      var e;
      (e = t.value) == null || e.setData(n.value);
    }), i({
      component: t
    }), (e, c) => (_(), h("div", k));
  }
});
export {
  E as VisDonutSelectors,
  L as default
};
